### ISSUE TITLE
Exception during creation of Credit Memo when tax and discount applies to shipping price

### ISSUE THREAD

**Issue by datasolution75** [magento/magento2#19429]
### Preconditions (*)

1. Magento 2 2.2.6

### Steps to reproduce (*)

1.  Make an Order with shipping cost and VAT on the shipping. (in my exemple 7.9€ with a 20% VAT rate)
2. Use a discount fixed on cart with an amount that can apply to shipping also (in my exemple 5€)

![image](https://user-images.githubusercontent.com/39830829/49149374-350bfa00-f30a-11e8-9d3e-019c1a05b5a4.png)

3. Invoice and then create a credit memo from the invoice.

### Expected result (*)

1. The credit memo page appear

### Actual result (*)

1. Exception Maximum shipping amount allowed to refund is: 7.89

> {"0":"Maximum shipping amount allowed to refund is: 7.89","1":"
> #0 HOME/vendor/magento/module-sales/Model/Order/Creditmemo.php(347): Magento\Sales\Model\Order\Creditmemo\Total\Shipping->collect(Object(Magento\Sales\Model\Order\Creditmemo\Interceptor))
> #1 HOME/vendor/magento/module-sales/Model/Order/CreditmemoFactory.php(132): Magento\Sales\Model\Order\Creditmemo->collectTotals()
> #2 HOME/vendor/magento/module-sales/Controller/Adminhtml/Order/CreditmemoLoader.php(209): Magento\Sales\Model\Order\CreditmemoFactory->createByInvoice(Object(Magento\Sales\Model\Order\Invoice), Array)
> #3 HOME/vendor/magento/module-sales/Controller/Adminhtml/Order/Creditmemo/NewAction.php(63): Magento\Sales\Controller\Adminhtml\Order\CreditmemoLoader->load()
> #4 HOME/vendor/magento/framework/App/Action/Action.php(107): Magento\Sales\Controller\Adminhtml\Order\Creditmemo\ewAction->execute()
> #5 HOME/vendor/magento/module-backend/App/AbstractAction.php(235): Magento\Framework\App\Action\Action->dispatch(Object(Magento\Framework\App\Request\Http))
> #6 HOME/vendor/magento/framework/Interception/Interceptor.php(58): Magento\Backend\App\AbstractAction->dispatch(Object(Magento\Framework\App\Request\Http))
> #7 HOME/vendor/magento/framework/Interception/Interceptor.php(138): Magento\Sales\Controller\Adminhtml\Order\Creditmemo\ewAction\Interceptor->___callParent('dispatch', Array)
> #8 HOME/vendor/magento/module-backend/App/Action/Plugin/Authentication.php(143): Magento\Sales\Controller\Adminhtml\Order\Creditmemo\ewAction\Interceptor->Magento\Framework\Interception\{closure}(Object(Magento\Framework\App\Request\Http))
> #9 HOME/vendor/magento/framework/Interception/Interceptor.php(135): Magento\Backend\App\Action\Plugin\Authentication->aroundDispatch(Object(Magento\Sales\Controller\Adminhtml\Order\Creditmemo\ewAction\Interceptor), Object(Closure), Object(Magento\Framework\App\Request\Http))
> #10 HOME/vendor/magento/framework/Interception/Interceptor.php(153): Magento\Sales\Controller\Adminhtml\Order\Creditmemo\ewAction\Interceptor->Magento\Framework\Interception\{closure}(Object(Magento\Framework\App\Request\Http))
> #11 HOME/generated/code/Magento/Sales/Controller/Adminhtml/Order/Creditmemo/NewAction/Interceptor.php(26): Magento\Sales\Controller\Adminhtml\Order\Creditmemo\ewAction\Interceptor->___callPlugins('dispatch', Array, NULL)
> #12 HOME/vendor/magento/framework/App/FrontController.php(55): Magento\Sales\Controller\Adminhtml\Order\Creditmemo\ewAction\Interceptor->dispatch(Object(Magento\Framework\App\Request\Http))
> #13 HOME/vendor/magento/framework/Interception/Interceptor.php(58): Magento\Framework\App\FrontController->dispatch(Object(Magento\Framework\App\Request\Http))
> #14 HOME/vendor/magento/framework/Interception/Interceptor.php(138): Magento\Framework\App\FrontController\Interceptor->___callParent('dispatch', Array)
> #15 HOME/vendor/magento/framework/Interception/Interceptor.php(153): Magento\Framework\App\FrontController\Interceptor->Magento\Framework\Interception\{closure}(Object(Magento\Framework\App\Request\Http))
> #16 HOME/generated/code/Magento/Framework/App/FrontController/Interceptor.php(26): Magento\Framework\App\FrontController\Interceptor->___callPlugins('dispatch', Array, Array)
> #17 HOME/vendor/magento/framework/App/Http.php(135): Magento\Framework\App\FrontController\Interceptor->dispatch(Object(Magento\Framework\App\Request\Http))
> #18 HOME/vendor/magento/framework/App/Bootstrap.php(257): Magento\Framework\App\Http->launch()
> #19 HOME/index.php(39): Magento\Framework\App\Bootstrap->run(Object(Magento\Framework\App\Http\Interceptor))
> #20 {main}","url":"/admin/sales/order_creditmemo/new/key/89b8c3c6e96cdcc5aa7981498fafe687026e77e2c9f20204bc5a2c723c4da936/order_id/112/invoice_id/106/","script_name":"/index.php"}

The "collect" function of Magento\Sales\Model\Order\Creditmemo\Total\Shipping should add the ShippingDiscountTaxCompensationAmount in the allowedTaxAmount variable.

Code is now : 

```
$allowedTaxAmount = $order->getShippingTaxAmount() - $order->getShippingTaxRefunded();
$baseAllowedTaxAmount = $order->getBaseShippingTaxAmount() - $order->getBaseShippingTaxRefunded();
```

It should be : 

```
$allowedTaxAmount = $order->getShippingTaxAmount() + $order->getShippingDiscountTaxCompensationAmount() - $order->getShippingTaxRefunded();
$baseAllowedTaxAmount = $order->getBaseShippingTaxAmount() + $order->getBaseShippingDiscountTaxCompensationAmnt() - $order->getBaseShippingTaxRefunded();
```

**Comment by magento-engcom-team** [comment]
Hi @datasolution75. Thank you for your report. 
To help us process this issue please make sure that you provided the following information:
- [ ] Summary of the issue
- [ ] Information on your environment
- [ ] Steps to reproduce
- [ ] Expected and actual results

Please make sure that the issue is reproducible on the vanilla Magento instance following **Steps to reproduce**. To deploy vanilla Magento instance on our environment, please, add a comment to the issue: 

`@magento-engcom-team give me $VERSION instance` 

where `$VERSION` is [version tags](https://github.com/magento/magento2/tags) (starting from 2.2.0+) or develop branches (for example: 2.3-develop). 
For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation.

@datasolution75 do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
- [ ] yes
- [ ] no

**Comment by magento-engcom-team** [comment]
Hi @engcom-backlog-andrii. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [x] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `Issue: Format is valid` will be added to the issue automatically. Please, edit issue description if needed, until label `Issue: Format is valid` appears.</details>

- [x] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `Issue: Clear Description` label to the issue by yourself.

- [x] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [x] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [x] 6. Add label `Issue: Confirmed` once verification is complete. 

- [x] 7. Make sure that automatic system confirms that report has been added to the backlog.

**Comment by orlangur** [comment]
@datasolution75 for suggesting code changes it is better to create a PR 😉

**Comment by datasolution75** [comment]
Sorry i'm new to this, what is a PR ?
And where can i create one ?

**Comment by ghost** [comment]
Hi @datasolution75. Thanks for the report. I was not able to reproduce this issue by following the steps you provided. My Grand Total (Excl tax.) is 41,94. I got this result: 
![item](https://user-images.githubusercontent.com/41946904/49160064-95149780-f32e-11e8-8816-c1584870a39d.png)

**Comment by datasolution75** [comment]
Hi, i see that you don't have exacty the same prices, and i think the shippings you set are not taxed.

All the products and the shipping method have a 20% tax rate. The prices are all including tax in the BO.
Here is the Tax config : 

![image](https://user-images.githubusercontent.com/39830829/49161673-9e4f3580-f329-11e8-8cad-07443e0ef556.png)

![image](https://user-images.githubusercontent.com/39830829/49161769-cb9be380-f329-11e8-9d68-5fb86486f22a.png)


Firt product price including tax is : 19.95
Second is 26.90
Shipping is 7.90.

All have taxes.

Here is the invoice i have in the database : 

![image](https://user-images.githubusercontent.com/39830829/49161200-a490e200-f328-11e8-8741-ee288d32f556.png)

![image](https://user-images.githubusercontent.com/39830829/49161238-becac000-f328-11e8-8f76-c2ba0f16fa8c.png)

**Comment by orlangur** [comment]
@datasolution75 please check https://github.com/magento/magento2/blob/2.3-develop/.github/CONTRIBUTING.md on how to propose some changes.

**Comment by magento-engcom-team** [comment]
@engcom-backlog-andrii Thank you for verifying the issue. Based on the provided information internal tickets `MAGETWO-96738` were created

**Comment by milindsingh** [comment]
@datasolution75 @orlangur can I create a PR with suggested changes ?

**Comment by datasolution75** [comment]
I don't have any objection on my side.

**Comment by vikalps4** [comment]
I am working on this issue at #dmdcindia1

**Comment by vikalps4** [comment]
I am working on this at #dmcdindia1

**Comment by mekedron** [comment]
@magento-engcom-team 
I'm pretty sure the field `base_shipping_discount_tax_compensation_amnt` will never have a value instead of null inside the `quote_address` table.

At first, take a look at this WONDERFUL phpdoc:
`\Magento\Quote\Model\Quote\Address` at line 84-85.
`setBaseShippingDiscountTaxCompensationAmnt`

Please, follow up this trace:
`\Magento\Quote\Model\Quote\TotalsCollector::collect` - line 148
`\Magento\Quote\Model\Quote\TotalsCollector::collectAddressTotals` - line 274
`\Magento\Tax\Model\Sales\Total\Quote\Tax::collect` - line 126
`\Magento\Tax\Model\Sales\Total\Quote\Shipping::collect` - line 56
`\Magento\Tax\Model\Sales\Total\Quote\CommonTaxCollector::processShippingTaxInfo` - line 743
`\Magento\Quote\Model\Quote\Address\Total::setBaseTotalAmount` - pay your attention to '_amont' literal.
`\Magento\Quote\Model\Quote\TotalsCollector::collectAddressTotals` - line 286

How should this field convert from `base_shipping_discount_tax_compensation_amount` into `base_shipping_discount_tax_compensation_amnt`?

**Comment by engcom-Charlie** [comment]
Hello @datasolution75 

I am not able to reproduce this issue on a fresh Magento 2.4-develop.

**Testing scenario:**

1. Make an Order with shipping cost and VAT on the shipping. (in my exemple 5$ with a 20% VAT rate)
2. Use a discount fixed on cart with an amount that can apply to shipping also (in my example 5$)

**Result:**
Credit memo has been created
![image](https://user-images.githubusercontent.com/51680850/71473409-90dfd580-27df-11ea-9abf-6927516fc14c.png)


So i have to close this issue as not reproducible.

Please feel free to comment, reopen or create new ticket according to the [Issue reporting guidelines](https://github.com/magento/magento2/wiki/Issue-reporting-guidelines) 
if you are still facing this issue on the latest Magento 2.4-develop branch. 

Thank you for collaboration.

**Comment by mekedron** [comment]
@engcom-Charlie please check [the comment](https://github.com/magento/magento2/issues/19429#issuecomment-564657965) I've left above
Is it related to this issue or I need to create another one?

**Comment by engcom-Charlie** [comment]
Hello @durexlovesex, please create a new issue for it.
Thanks for contributing!
